sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/HashPartitionPushDownSupport.java

```diff
@@ -0,0 +1,30 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.sources.v2.reader;
+
+/**
+ * A mix in interface for `DataSourceV2Reader`. Users can implement this interface to pre-partition
+ * the data and avoid shuffle at Spark side.
+ */
+public interface HashPartitionPushDownSupport {
```

Comment:
Actually it's designed for case 1...

At the beginning, I was thinking of doing hash partitioning propagation, which looks more natural to support case 1. However, some data sources may be partitioned in multiple ways, e.g. it's hash partitioned by a, b, and also hash partitioned by c, d. It will be a little messy to ask a data source to propagate all the possible hash partitionings.

Instead, I think it can be simpler if we do hash partitioning push down. The implementation just need to tell us if it can do this hash partitioning or not. That said, this method should return a boolean instead of void.